---
title: "PATRIC_Wrangling"
author: "Rachel Xu"
date: "12/8/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(dplyr)
path <- "/Users/rachel/Dropbox/5.Rachel-projects/Phylogeography/SeroAll/"


```


```{r}

all_data_table <- read.csv(paste0(path, "PATRIC_genome_all_732.csv")) %>% select(c("Genome.ID", "Genome.Name", "Organism.Name", "Serovar", "BioProject.Accession", "BioSample.Accession", "Assembly.Accession", "SRA.Accession", "GenBank.Accessions", "Sequencing.Centers", "Sequencing.Platform", "Sequencing.Depth", "PATRIC.CDS", "GC.Content", "Isolation.Source","Isolation.Country", "Geographic.Location", "Host.Name", "Collection.Date"))


pathogenic <- c("interrogans","mayottensis","santarosai","borgpetersenii","dzianensis","alstonii", "adleri","noguchii", "kirschneri", "barantonii", "weilii","putramalaysiae","alexanderi","gomenensis","kmetyi","yasudae","elllisii","adleri","stimsonii","kmetyi", "alstoni")

intermediate <- c("andrefontaineae","wolffii","sarikeiensis","fainei","selangorensis","inadai","dzoumogneensis","fluminis","licerasiae","koniamboensis","fletcheri","langatensis","semungkisensis","perolatii","neocaledonica","haakeii","saintgironsiae","hartskeerlii","johnsonii","venezuelensis", "broomii")

saprophytic <- c("meyeri", "idonii", "wolbachii","biflexa", "levettii", "brenneri", "ilyithenensis", "bandrabouensis", "bouyouniensis", "mtsangambouensis", "jelokensis", "congkakensis", "kemamanensis", "ognonensis", "perdikensis", "montravelensis", "bourretii", "noumeaensis", "kanakyensis", "vanthielii")
   
# label all lepto genomes from PATRIC with phylogenetic group              
all_data_table <- all_data_table %>% 
  mutate(species = unlist(lapply(strsplit(Genome.Name, " "),function(item) {unlist(item)[2]}))) %>%
  subset(species != "illini") %>% 
  mutate(group=ifelse(species %in% pathogenic, "pathogenic", 
                      ifelse(species %in% intermediate, "intermediate", 
                             ifelse(species %in% saprophytic, "saprophytic", "unknown")))) %>%
  subset(group == "pathogenic") 

# get all isolates with serotype metadata available 
patric_sero <- all_data_table %>% 
  subset(Serovar != "" & Serovar != "serovar undetermined" & Serovar != "serovar s" & Serovar != "Room22") %>%
  mutate(Serovar_2 = ifelse(grepl("serovar",Serovar,fixed = TRUE), unlist(lapply(Serovar,function(x) {unlist(strsplit(x," "))[match("serovar", unlist(strsplit(x," ")))+1]})), Serovar)) %>% 
  select(c("Genome.Name", "BioProject.Accession" , "BioSample.Accession", "Sequencing.Centers", "Host.Name", "Isolation.Country", "Geographic.Location", "species", "Serovar_2","Collection.Date"))

patric_sero
```

## To select all pathogenic isolates from NCBI+Self assembled genomes
   - add metadata to those genomes with python script **all_lepto_metadata.ipynb**
   - keep working after read in the metadata file
```{r}
# 
# # get all pathogenic isolates from self-assembled + NCBI biosamples
# NCBI_Self <- read.csv("/Users/rachel/Dropbox/5.Rachel-projects/Phylogeography/Lepto_assemblies_V2/ncbi+self_973.csv") %>% 
#   subset(unlist(lapply(strsplit(species, " "),function(item) {unlist(item)[2]})) %in% pathogenic)
# # write.csv(NCBI_Self, paste0(path,"self_assembled_pathogenic_782.csv"), sep=",", row.names = FALSE)

# to get metadata for all pagthogenic lepto 
# with python script all_lepto_metadata.ipynb
ncbi_self_sero <- read.csv(paste0(path, "ncbiself_metadata_pathogenic_782.csv")) %>%
  mutate(species = unlist(lapply(biosample_name, function(item){unlist(strsplit(item, " "))[2]}))) %>%
  mutate(Serovar_2=ifelse(grepl("serovar",biosample_name,fixed = TRUE), unlist(lapply(biosample_name,function(x) {unlist(strsplit(x," "))[match("serovar", unlist(strsplit(x," ")))+1]})), Serovar))  %>%
  subset(Serovar_2 != "" & Serovar_2 != "Unknown") %>%
  select(c("BioSample.Accession", "biosample_name", "institute", "Bioproject", "Collection.Date","Host.Name", "Geographic.Location","country", "Serovar_2", "species" ))

colnames(ncbi_self_sero) <- c("BioSample.Accession","Genome.Name", "Sequencing.Centers", "BioProject.Accession","Collection.Date", "Host.Name", "Geographic.Location", "Isolation.Country", "Serovar_2", "species" )

# all serovar to lower case
ncbi_self_sero$Serovar_2 <- tolower(ncbi_self_sero$Serovar_2)

all_pathogenic_sero <- full_join(ncbi_self_sero, patric_sero, by="BioSample.Accession")# %>% select(c("BioSample.Accession"))
# write.csv(all_pathogenic_sero, paste0(path, "all_pathogenic_sero.csv"))
# "all_pathogenic_sero.csv" is manually fixed by combining metadata for both database


```

# Serovar to serogroup dict
```{r}
Anarum <- c("Pingchang")
Andaman <- c("Andamana")
Australis <- c("Australis","Bajan","Bangkok","Bratislava","Fugis","Hawain","Jalna","Lora","Muenchen","Nicaragua","Peruviana","Pina","Ramisi","Rushan","Soteropolitana")
Autumnalis <- c("Alice","Autumnalis","Bangkinang","Bim","Bulgarica","Butembo","Carlos","Erinaceiauriti","Fortbragg","Lambwe","Mooris","Mujunkumi","Nanla","Rachmati","Srebarna","Sumatrana","Weerasinghe")
Ballum <- c("Aroborea","Ballum","Castellonis","Guangdong","Kenya","Peru","Sichuan")
Bataviae <- c("Argentiniensis","Balboa","Bataviae","Brasiliensis","Claytoni","Djatzi","Kobbe","Losbanos","Paidjan","Rioja","Santarosa","Waimea")
Canicola <- c("Bafani","Benjamini","Bindjei","Broomi","Canicola","Galtoni","Jonsis","Kamituga","Kuwait","Malaya","Manilae","Portlandvere","Qunjian","Schueffneri","Sumneri")
Celledoni <- c("Anhoa","Celledoni","Cynopteri","Hainan","Mengdeng","Whitcombi")
Codice<- c("Codice")
Cynopteri <- c("Cynopteri","Tingomaria")
Djasiman <- c("Agogo","Buenos Aires","Djasiman","Gurungi","Huallaga","Sentot")
Grippotyphosa <- c("Bananal","Canalzonae","Dadas","Grippotyphosa","Grippotyphosa Type Duyster","Grippotyphosa Type Moskva","Honghe","Huanuco","Liangguang","Linhai","Muelleri","Ratnapura","Valbuzzi","Vanderhoedeni")
Hebdomanis <- c("Borincana","Goiano","Hebdomadis","Jules","Kabura","Kambale","Kremastos","Manzhuang","Maru","Medanensis","Nanding","Nona","Sanmartini","Worsfoldi")
Holland <- c("Holland")
Hurstbridge <- c("Hurstbridge","Hurstbridge type HB6")
Icterohaemorrhagiae <- c("Birkini","Bogvere","Copenhageni","Dakota","Gem","Hongchon","Honghe","Hualin","Icterohaemorrhagiae","Lai","Lai Type Langkawi","Liangshan","Mankarso","Monymusk","Mwogolo","Naam","Nanxi","Ndahambukuje","Ndambari","Smithi","Sokoine","Tonkini","Yeoncheon", "copenhageni/icterohaemorrhagiae")
Iquitos <- c("Varillal")
Javanica <- c("Arenal","Ceylonica","Coxi","Dehong","Fluminense","Javanica","Kalimantani","Mengla","Mengma","Mengrum","Menoni","Menrun","Poi","Sofia","Sorexjalna","Vargonicas","Yaan","Zhenkang")
Louisiana <- c("Lanka","Louisiana","Orleans")
Lyme <- c("Lyme")
Manhao <- c("Lichuan","Lincang","Lushui","Manhao","Qingshui")
Mini <- c("Beye","Georgia","Hekou","Mini","Perameles","Ruparupae","Swajizak","Tabaquite","Yunnan")
Panama <- c("Cristobali","Mangus","Panama")
Pomona <- c("Altodouro","Kennewicki","Kunming","Mozdok","Pomona","Proechimys","Tropica","Tsaratsovo")
Pyrogenes <- c("Abramis","Alexi","Biggis","Camlo","Costa Rica","Guaratuba","Hamptoni","Kwale","Manilae","Menglian","Myocastoris","Nigeria","Princestown","Pyrogenes","Robinsoni","Varela","Zanoni")
Ranarum <- c("Evansi","Pingchang","Ranarum")
Sarmin <- c("Cuica","Machiguenga","Rio","Sarmin","Waskurin","Weaveri")
Sehgali <- c("Portblairi")
Sejroe <- c("Balcanica","Caribe","Dikkeni","Geyaweera","Gorgas","Guaricura","Haemolytica","Hardjo","Hardjo type Bovis","Hardjo type Prajitno",
            "hardjo-bovis",
            "Hardjobovis",
            "hardjo-prajitno","Istrica","Jin","Medanensis","Nyanza","Polonica","Recreo","Ricardi","Roumanica","Saxkoebing","Sejroe","Trinidad","Unipertama","Wolffi")
Semaranga <- c("Patoc","Semaranga","Saopaulo")
Shermani <- c("Aguaruna","Badudieri","Carimagua","Luis","Shermani")
Tarassovi <- c("Atchafalaya","Atlantae","Bakeri","Banna"," Bravo","Chagres","Corredores","Darien","Gatuni","Gengma","Guidae","Kanana","Kaup","Kisuba","Langati","Malaysia","Mengpeng","Mogden","Moldaviae","Navet","Ningxia","Rama","Sulzerae","Tarassovi","Topaz","Tunis","Vughia","Yunxian")
Szwajizak <- c("szwajizak")
Arborea <- c("Arborea")
Heyan <- c("heyan")

serogroup_dict <- list(Anarum = Anarum,Andaman=Andaman, Australis=Australis,Autumnalis=Autumnalis,Ballum=Ballum,Bataviae=Bataviae, Canicola=Canicola,Celledoni=Celledoni,Codice=Codice,Cynopteri=Cynopteri,Djasiman=Djasiman,Grippotyphosa=Grippotyphosa,Hebdomanis=Hebdomanis,Holland=Holland,Hurstbridge=Hurstbridge,Icterohaemorrhagiae=Icterohaemorrhagiae,Iquitos=Iquitos,Javanica=Javanica,Louisiana=Louisiana,Lyme=Lyme,Manhao=Manhao,Mini=Mini,Panama=Panama,Pomona=Pomona,Pyrogenes=Pyrogenes,Ranarum=Ranarum,Sarmin=Sarmin,Sehgali=Sehgali,Sejroe=Sejroe,Semaranga=Semaranga,Shermani=Shermani,Tarassovi=Tarassovi,Szwajizak=Szwajizak, Arborea=Arborea,  Heyan = Heyan)
serogroup_dict <- lapply(serogroup_dict, tolower)
```

# add serogroup to metadata 
```{r}

pathogenic_sero_db <- read.csv(paste0(path, "all_pathogenic_sero_597.csv"))
serovars <- pathogenic_sero_db$Serovar_2.x

serogroups <- c()
j = 0
for(item in serovars){
  j=j+1
  i=0
  in_dict <- 0
  item <- tolower(item)
  # print(paste0("serovar: ",item))
  for(serogroup in serogroup_dict){
    i=i+1
    if(item %in% serogroup){
      in_dict <- 1
      # print(paste0("serogroup: ",names(serogroup_dict)[i]))
      serogroups <- c(serogroups,names(serogroup_dict)[i])
      break
    }
  }
  if(in_dict == 0){
    # print("not_in_dict")
    serogroups <- c(serogroups,"not_in_dict")
  }
}

length(serogroups)
# serovars
pathogenic_sero_db$Serovar_2.x <- unlist(lapply(pathogenic_sero_db$Serovar_2.x, function(x){
  ifelse(x=="Hardjobovis","hardjo-bovis",ifelse(x=="copenhageni/icterohaemorrhagiae","copenhageni",tolower(x)))
}[1]))
pathogenic_sero_db$serogroups <-serogroups

# write.csv(pathogenic_sero_db,paste0(path,"all_pathogenic_sero_597.csv"))
```

# make the host names consistent
```{r}

pathogenic_sero_db <- read.csv(paste0(path,"all_pathogenic_sero_597.csv"))



pathogenic_sero_db_2 <-pathogenic_sero_db %>% mutate(Host_2 = unlist(lapply(Host.Name.x, function(x){
  ifelse(grepl(",",x), unlist(strsplit(x,", ", fixed = TRUE))[2],
         ifelse(x=="",NA,ifelse(x=="Unknown",NA,
                ifelse(x=="dog", "Canis lupus familiaris", 
                       ifelse(x=="rat", "Rattus sp.", 
                              ifelse(x=="rodent","Rodentia",
                                     ifelse(grepl("caballus",x), "Equus ferus caballus", 
                                            ifelse(tolower(x)=="frog", "Anura", 
                                                   ifelse(x=="fox", "Vulpes sp.",x)))))))))
})))

unique(pathogenic_sero_db_2$Host_2)

# write.csv(pathogenic_sero_db_2,paste0(path,"all_pathogenic_sero_597.csv"),row.names = FALSE)
```


#Analyze all pathogenic Lepto with sero metadata
```{r}

library(ggplot2)

sero_pathogenic <- read.csv(paste0(path,"all_pathogenic_sero_597.csv"))

length(unique(sero_pathogenic$Serovar_2.x))
ggplot(sero_pathogenic, aes(x=Serovar_2.x)) +
  geom_bar() +
  theme(axis.text.x= element_text(angle=90))

length(unique(sero_pathogenic$serogroups))
ggplot(sero_pathogenic, aes(x=serogroups)) +
  geom_bar() +
  theme_bw() +
  theme(axis.text.x= element_text(angle=90)) 

length(unique(sero_pathogenic$Host.Name.x))
unique(sero_pathogenic$Isolation.Country.x)
```