---
title: "all_leptopsira_WGS"
author: "Rachel Xu"
date: "1/14/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r}
library(dplyr)
library(magrittr)
path <- "/Users/rachel/Dropbox/5.Rachel-projects/Phylogeography/Lepto_information_V3/all_species/assemblies/"

```

# Read in previously organized metadata files
- merged manually to take metadata from all databases
```{r}
# # all patric wgs metadata 1/14/2021
# patric_all <- read.csv(paste0(path,"all_patric_wgs.csv")) %>% select(c("Genome.Name",	"Serovar",		"BioSample.Accession",	"Sequencing.Centers",	"Sequencing.Platform",	"Sequencing.Depth",	"GC.Content",	"Isolation.Source",	"Isolation.Country",	"Geographic.Location",	"Host.Name",	"Collection.Date",	"species",	"group",))
# 
# # current ncbi data I have extracted, but not the newly added ones (1/14/2021)
# ncbi_current <- read.csv(paste0(path, "ncbi+self_973.csv")) %>% 
#   rename(BioSample.Accession = Sample)
# ncbi_new_meta_search <- read.csv(paste0(path, "biosample_metadata_973.csv"))
# merged_ncbi <- full_join(ncbi_current, ncbi_new_meta_search, by="BioSample.Accession")
# 
# # read in sapelo2 samples that were not in V2 assemblies' metadata
# sapelo2_extra <- read.csv(paste0(path, "biosample_metadata_sapelo.csv"))
# 
# # merge metadata together, also to see which biosample still need metadata 
# merged_metadata <- full_join(patric_all, merged_ncbi, by= "BioSample.Accession")
# merged_metadata <- full_join(merged_metadata, sapelo2_extra, by ="BioSample.Accession")


# ****DO NOT OVERRIDE THIS FILE< MANUALLY MERGED
# write.csv(merged_metadata, paste0(path, "NCBI_patric_all_wgs_metadata.csv"))

```


# wrangling with merged metadata 

# 1) need parse out species
```{r}

all_asm_meta <- read.csv(paste0(path, "NCBI_patric_all_wgs_metadata.csv"))

# 1) need parse out species
new_species <- unlist(lapply(all_asm_meta$Genome.Name.x, function(X) {unlist(strsplit(X, " " ))[2]}))
all_asm_meta$species <- new_species
all_asm_meta <- all_asm_meta %>% select(-c("species.x"))









```

# 2) need to put species into groups
```{r}

# 2) need to put species into groups

pathogenic <- c("interrogans","mayottensis","santarosai","borgpetersenii","dzianensis","alstonii", "adleri","noguchii", "kirschneri", "barantonii", "weilii","putramalaysiae","alexanderi","gomenensis","kmetyi","yasudae","elllisii","adleri","stimsonii","kmetyi", "alstoni")

intermediate <- c("andrefontaineae","wolffii","sarikeiensis","fainei","selangorensis","inadai","dzoumogneensis","fluminis","licerasiae","koniamboensis","fletcheri","langatensis","semungkisensis","perolatii","neocaledonica","haakeii","saintgironsiae","hartskeerlii","johnsonii","venezuelensis", "broomii")

saprophytic <- c("meyeri", "idonii", "wolbachii","biflexa", "levettii", "brenneri", "ilyithenensis", "bandrabouensis", "bouyouniensis", "mtsangambouensis", "jelokensis", "congkakensis", "kemamanensis", "ognonensis", "perdikensis", "montravelensis", "bourretii", "noumeaensis", "kanakyensis", "vanthielii","ellinghausenii", "kobayashii", "harrisiae", "ryugenii")

# label all lepto genomes from PATRIC with phylogenetic group              
all_asm_meta <- all_asm_meta %>% 
  subset(species != "illini") %>% 
  mutate(group=ifelse(species %in% pathogenic, "pathogenic", 
                      ifelse(species %in% intermediate, "intermediate", 
                             ifelse(species %in% saprophytic, "saprophytic", "unknown")))) 


# all_asm_meta %>% subset(group == "unknown")

```


# parse serovars
```{r}

all_asm_meta <- all_asm_meta %>%
  mutate(Serovar=tolower(ifelse(grepl("serovar",Genome.Name.x,fixed = TRUE), unlist(lapply(Genome.Name.x,function(x) {unlist(strsplit(x," "))[match("serovar", unlist(strsplit(x," ")))+1]})), Serovar.x)))  

all_asm_meta <- all_asm_meta %>% select(-c("Serovar.x"))

```

### Serogroup dictionary
```{r}

Anarum <- c("Pingchang")
Andaman <- c("Andamana")
Australis <- c("Australis","Bajan","Bangkok","Bratislava","Fugis","Hawain","Jalna","Lora","Muenchen","Nicaragua","Peruviana","Pina","Ramisi","Rushan","Soteropolitana")
Autumnalis <- c("Alice","Autumnalis","Bangkinang","Bim","Bulgarica","Butembo","Carlos","Erinaceiauriti","Fortbragg","Lambwe","Mooris","Mujunkumi","Nanla","Rachmati","Srebarna","Sumatrana","Weerasinghe")
Ballum <- c("Aroborea","Ballum","Castellonis","Guangdong","Kenya","Peru","Sichuan")
Bataviae <- c("Argentiniensis","Balboa","Bataviae","Brasiliensis","Claytoni","Djatzi","Kobbe","Losbanos","Paidjan","Rioja","Santarosa","Waimea")
Canicola <- c("Bafani","Benjamini","Bindjei","Broomi","Canicola","Galtoni","Jonsis","Kamituga","Kuwait","Malaya","Manilae","Portlandvere","Qunjian","Schueffneri","Sumneri")
Celledoni <- c("Anhoa","Celledoni","Cynopteri","Hainan","Mengdeng","Whitcombi")
Codice<- c("Codice")
Cynopteri <- c("Cynopteri","Tingomaria")
Djasiman <- c("Agogo","Buenos Aires","Djasiman","Gurungi","Huallaga","Sentot")
Grippotyphosa <- c("Bananal","Canalzonae","Dadas","Grippotyphosa","Grippotyphosa Type Duyster","Grippotyphosa Type Moskva","Honghe","Huanuco","Liangguang","Linhai","Muelleri","Ratnapura","Valbuzzi","Vanderhoedeni")
Hebdomanis <- c("Borincana","Goiano","Hebdomadis","Jules","Kabura","Kambale","Kremastos","Manzhuang","Maru","Medanensis","Nanding","Nona","Sanmartini","Worsfoldi")
Holland <- c("Holland")
Hurstbridge <- c("Hurstbridge","Hurstbridge type HB6")
Icterohaemorrhagiae <- c("Birkini","Bogvere","Copenhageni","Dakota","Gem","Hongchon","Honghe","Hualin","Icterohaemorrhagiae","Lai","Lai Type Langkawi","Liangshan","Mankarso","Monymusk","Mwogolo","Naam","Nanxi","Ndahambukuje","Ndambari","Smithi","Sokoine","Tonkini","Yeoncheon", "copenhageni/icterohaemorrhagiae")
Iquitos <- c("Varillal")
Javanica <- c("Arenal","Ceylonica","Coxi","Dehong","Fluminense","Javanica","Kalimantani","Mengla","Mengma","Mengrum","Menoni","Menrun","Poi","Sofia","Sorexjalna","Vargonicas","Yaan","Zhenkang")
Louisiana <- c("Lanka","Louisiana","Orleans")
Lyme <- c("Lyme")
Manhao <- c("Lichuan","Lincang","Lushui","Manhao","Qingshui")
Mini <- c("Beye","Georgia","Hekou","Mini","Perameles","Ruparupae","Swajizak","Tabaquite","Yunnan")
Panama <- c("Cristobali","Mangus","Panama")
Pomona <- c("Altodouro","Kennewicki","Kunming","Mozdok","Pomona","Proechimys","Tropica","Tsaratsovo")
Pyrogenes <- c("Abramis","Alexi","Biggis","Camlo","Costa Rica","Guaratuba","Hamptoni","Kwale","Manilae","Menglian","Myocastoris","Nigeria","Princestown","Pyrogenes","Robinsoni","Varela","Zanoni")
Ranarum <- c("Evansi","Pingchang","Ranarum")
Sarmin <- c("Cuica","Machiguenga","Rio","Sarmin","Waskurin","Weaveri")
Sehgali <- c("Portblairi")
Sejroe <- c("Balcanica","Caribe","Dikkeni","Geyaweera","Gorgas","Guaricura","Haemolytica","Hardjo","Hardjo type Bovis","Hardjo type Prajitno",
            "hardjo-bovis",
            "Hardjobovis",
            "hardjo-prajitno","Istrica","Jin","Medanensis","Nyanza","Polonica","Recreo","Ricardi","Roumanica","Saxkoebing","Sejroe","Trinidad","Unipertama","Wolffi")
Semaranga <- c("Patoc","Semaranga","Saopaulo")
Shermani <- c("Aguaruna","Badudieri","Carimagua","Luis","Shermani")
Tarassovi <- c("Atchafalaya","Atlantae","Bakeri","Banna"," Bravo","Chagres","Corredores","Darien","Gatuni","Gengma","Guidae","Kanana","Kaup","Kisuba","Langati","Malaysia","Mengpeng","Mogden","Moldaviae","Navet","Ningxia","Rama","Sulzerae","Tarassovi","Topaz","Tunis","Vughia","Yunxian")
Szwajizak <- c("szwajizak")
Arborea <- c("Arborea")
Heyan <- c("heyan")

serogroup_dict <- list(Anarum = Anarum,Andaman=Andaman, Australis=Australis,Autumnalis=Autumnalis,Ballum=Ballum,Bataviae=Bataviae, Canicola=Canicola,Celledoni=Celledoni,Codice=Codice,Cynopteri=Cynopteri,Djasiman=Djasiman,Grippotyphosa=Grippotyphosa,Hebdomanis=Hebdomanis,Holland=Holland,Hurstbridge=Hurstbridge,Icterohaemorrhagiae=Icterohaemorrhagiae,Iquitos=Iquitos,Javanica=Javanica,Louisiana=Louisiana,Lyme=Lyme,Manhao=Manhao,Mini=Mini,Panama=Panama,Pomona=Pomona,Pyrogenes=Pyrogenes,Ranarum=Ranarum,Sarmin=Sarmin,Sehgali=Sehgali,Sejroe=Sejroe,Semaranga=Semaranga,Shermani=Shermani,Tarassovi=Tarassovi,Szwajizak=Szwajizak, Arborea=Arborea,  Heyan = Heyan)
serogroup_dict <- lapply(serogroup_dict, tolower)


```


# 4ï¼‰assign serogroup base on serovar
```{r}
serovars <- all_asm_meta$Serovar 
serogroups <- c()
j = 0
for(item in serovars){
  j=j+1
  i=0
  in_dict <- 0
  item <- tolower(item)
  # print(paste0("serovar: ",item))
  for(serogroup in serogroup_dict){
    i=i+1
    if(item %in% serogroup){
      in_dict <- 1
      # print(paste0("serogroup: ",names(serogroup_dict)[i]))
      serogroups <- c(serogroups,names(serogroup_dict)[i])
      break
    }
  }
  if(in_dict == 0){
    # print("not_in_dict")
    serogroups <- c(serogroups,"not_in_dict")
  }
}

length(serogroups)


all_asm_meta$serogroups <-serogroups

all_asm_meta %<>% subset(BioSample.Accession != "SAMN09220899" & BioSample.Accession != "")


```

# 4) Parse Country
```{r}

all_asm_meta$Geographic.Location.x[is.na(all_asm_meta$Geographic.Location.x)] <- "NA"

location <- all_asm_meta$Geographic.Location.x

# take country info form format with "country : region" in location
country <- unlist(lapply(location, function(x) {
  ifelse(grepl(":", x, fixed = TRUE),
         unlist(strsplit(x, ":"))[1], x)
  }))

# take country info form format with "region , country" in location
country <- unlist(lapply(country, function(x) {
  ifelse(grepl(",", x, fixed = TRUE),
         unlist(strsplit(x, ","))[2], x)
  }))

country <- lapply(country, function(x){
if(x == "Pancas Portugal" | x == "Salvador Brazil" | x == "Mwembesongo Tanzania" | x == "Hawaii USA"){
  unlist(strsplit(x, " "))[2]
  }else if(x == "Elst Gelderland" | x== "Nieuwegein"){
    "Netherlands"
  }else if(x == "Belgian Congo" | x == "Zaire"){
    "Congo"
  }else if(x == "Linhai Zhejiang"){
    "China"
  }else if(x == "The Netherlands" | x == "Th eNetherlands"){
    "Netherlands"
  }else if(x == "NA" | x== "missing" | x == ""){
    "NA"
  }else{
    x
  }
})

# trim white space
country <- trimws(country)

all_asm_meta$country <- country
# DONT OVERRIDE, species, serovar, serogroup extracted
# write.csv(all_asm_meta, paste0(path, "NCBI_patric_all_wgs_metadata_2.csv"))
```

# 5) Overall parsing


```{r}
all_db <- read.csv(paste0(path,"NCBI_patric_all_wgs_metadata_2.csv"))

# unique(all_db$serogroups)
# unique(all_db$Collection.Date.x)

all_db$serogroups[all_db$serogroups == "not_in_dict"] <- "NA"
all_db$Collection.Date.x[all_db$Collection.Date.x == "" | all_db$Collection.Date.x == "missing"] <- "NA"

lepto_dated <- all_db %>% subset(Collection.Date.x != "NA")

# DO NOT OVERRIDE, DATE checked and format changed
# write.csv(lepto_dated, paste0(path, "all_lepto_dated.csv"))

```

### Species were mislabelled for some strains, will edited underneath
```{r}

all_lepto_dated <- read.csv(paste0(path, "all_lepto_dated.csv")) 

# two ways to do this 
all_lepto_dated <- within(all_lepto_dated,species[BioSample.Accession == "SAMN02436491"] <- "noguchii")

all_lepto_dated <- within(all_lepto_dated, {
    f <- BioSample.Accession == 'SAMN04488387' & species == 'interrogans'
    BioSample.Accession[f] <- 'SAMN04488387'
    species[f] <- 'kirschneri'
    group[f] <- ifelse(species[f] %in% pathogenic, "pathogenic", 
                      ifelse(species[f] %in% intermediate, "intermediate", 
                             ifelse(species[f] %in% saprophytic, "saprophytic", "unknown")))
})

all_lepto <- read.csv(paste0(path, "NCBI_patric_all_wgs_metadata_2.csv")) 

# two ways to do this 
all_lepto <- within(all_lepto,species[BioSample.Accession == "SAMN02436491"] <- "noguchii")

all_lepto <- within(all_lepto, {
    f <- BioSample.Accession == 'SAMN04488387' & species == 'interrogans'
    BioSample.Accession[f] <- 'SAMN04488387'
    species[f] <- 'kirschneri'
    group[f] <- ifelse(species[f] %in% pathogenic, "pathogenic", 
                      ifelse(species[f] %in% intermediate, "intermediate", 
                             ifelse(species[f] %in% saprophytic, "saprophytic", "unknown")))
}) %>%select(-c("X"))

# write into csv, do not override
# write.csv(all_lepto_dated, paste0(path, "all_lepto_dated.csv"))
# write.csv(all_lepto, paste0(path, "NCBI_patric_all_wgs_metadata_2.csv"))
```
### plot serotypes
```{r}
# all lepto info
sero_lepto <- read.csv(paste0(path,"NCBI_patric_all_wgs_metadata_2.csv")) %>% subset(group == "pathogenic") %>% subset(serogroups != "not_in_dict")

ggplot(sero_lepto, aes(x=serogroups))+
  geom_bar() +
  theme_bw() +
  labs(x="Serogroups",y="Number of Isolates")+
  theme(axis.text = element_text(angle=90,size = 12),axis.title = element_text(size=12))
```

## 5.1) plot groups
```{r}

library(ggplot2)

# number of isolates in each group
ggplot(all_db, aes(x=group)) +
  geom_bar() +
  geom_text(stat="count",aes(label=..count..,vjust=-0.5,degree=25),size = 4) + 
  theme_bw() +
  labs(title = "Number of Leptospira in Each Group")+
  theme(axis.text = element_text(size=12), axis.title = element_text(size = 12))

# Number of pathogenic isolates with serotype metadata
path_sero <- all_db %>% subset(group == "pathogenic") %>% subset(serogroups != "NA")
length(unique(path_sero$Serovar))
ggplot(path_sero, aes(x=serogroups)) +
  geom_bar() +
  geom_text(stat="count",aes(label=..count..,vjust=-0.5,degree=25),size = 4) + 
  theme_bw() +
  labs(title = "Number of Pathogenic Leptospira with Serotypes")+
  theme(axis.text = element_text(size=12, angle = 90), axis.title = element_text(size = 12))



# Number of pathogenic isolates with collection date metadata
path_dated <- all_db %>% subset(group == "pathogenic") %>% subset(Collection.Date.x != "NA")
length(path_dated$Collection.Date.x)
unique(path_dated$Collection.Date.x) 

path_dated_location <- path_dated %>% subset(country != "NA")
length(path_dated_location$country)
unique(path_dated_location$country)
table(path_dated_location$country)

```
# plot countries onto the map

- the codes below were adpated from this site:
https://slcladal.github.io/maps.html

```{r}
library(ggplot2)
library(sf)
library(rworldmap)
library(rnaturalearth) # data
library(rnaturalearthdata) #data
library(rgeos)

num_per_country <- as.data.frame(table(path_dated_location$country))
colnames(num_per_country) <- c("country", "number")
plot(getMap()[which(getMap()$ADMIN=='Mayotte'),] )
# num_per_country <- path_dated_location %>% select(c(country))

world <- ne_countries(scale = "large", returnclass = "sf")
join_country_map <- joinCountryData2Map(num_per_country, 
                                  joinCode = "NAME",
                                  nameJoinColumn = "country", verbose=T, mapResolution = "high")
pdf(paste0(path, "WGS_global_distrib_map.pdf"))
prevalentMap <- mapCountryData(join_country_map, 
                            nameColumnToPlot="number",
                            oceanCol = "azure2",
                            catMethod = "pretty",
                            missingCountryCol = gray(.8),
                            addLegend = F,
                            mapTitle = "",
                            border = NA)
# # add legend and display map
do.call(addMapLegendBoxes, c(prevalentMap,
                             x = 'bottomleft',
                             title = "Num. Leptospira genomes",
                             horiz = FALSE,
                             bg = "white",
                            cex=0.8))

dev.off()
```

```{r}

world <- ne_countries(scale = "large", returnclass = "sf")

merged <- merge(world, num_per_country, by = "country")

merged$number

worldMap <- ggplot(data = world, fill="gold") +
  geom_sf() +
  labs( x = "Longitude", y = "Latitude") +
  ggtitle("World map", subtitle = paste0("(", length(unique(world$admin)), " countries)")) 
  
worldMap
  

```

## 5.2) plot years
```{r}
library(ggplot2)

all_lepto_dated <- read.csv(paste0(path, "all_lepto_dated.csv"))

length(all_lepto_dated$BioSample.Accession)

# analyze temporal signal by species # interrogans

interrogans_path <- "/Users/rachel/Dropbox/5.Rachel-projects/Phylogeography/Lepto_information_V3/interrogans/"
interrogans_dated <- all_lepto_dated %>% subset(species == "interrogans" ) #%>% select(BioSample.Accession)
# write.csv(interrogans_dated, paste0(interrogans_path, "china_interrogans_acc_list.txt"), quote = FALSE, sep = "\n", col.names = NA, row.names = FALSE)
length(interrogans_dated$BioSample.Accession)

ggplot(interrogans_dated, aes(x=Host)) +
  geom_bar() +
  theme(axis.text.x = element_text(angle=90)) #+
    scale_x_continuous(breaks = seq(1910,2020,by=2))
    

```

### plot Asian Isolates
```{r}

asia_int_dated <- interrogans_dated %>% subset(country =="China") #%>% select(BioSample.Accession)

ggplot(asia_int_dated, aes(x=Year)) +
  geom_bar() +
  theme(axis.text.x = element_text(angle=90)) +
    scale_x_continuous(breaks = seq(1910,2020,by=2))
# write.csv(asia_int_dated, paste0(interrogans_path, "Asia_interrogans_acc_list.txt"), quote = FALSE, sep = "\n", col.names = NA, row.names = FALSE)

```

### Borgpetersenii NZ
```{r}

borg_path <- "/Users/rachel/Dropbox/5.Rachel-projects/Phylogeography/Lepto_information_V3/borgpetersenii/"
borg_dated <- all_lepto_dated %>% subset(species == "borgpetersenii" & country == "New Zealand") #%>% select(BioSample.Accession) 
borg_dated 
# # add host names to NZ isolates
# all_lepto_dated_1 <- all_lepto_dated %>% mutate(Host = ifelse(BioSample.Accession == "SAMN16295653" | BioSample.Accession == "SAMN16295652"| BioSample.Accession =="SAMN16295637", "Bos sp.",
#          ifelse(BioSample.Accession =="SAMN16295647" | BioSample.Accession =="SAMN16295648", "Cervus elaphus",
#                 ifelse( BioSample.Accession =="SAMN16295641"| BioSample.Accession =="SAMN16295638" | BioSample.Accession == "SAMN16295642" | BioSample.Accession =="SAMN16295639"| BioSample.Accession == "SAMN16295643"|BioSample.Accession == "SAMN16295640" ,"Ovis sp.", Host))))
# 
# all_lepto_dated_2 <- all_lepto_dated_1 %>%select(-c(Host.Name.y))
# write.csv(all_lepto_dated_1,paste0(path, "all_lepto_dated.csv"))
# write.csv(borg_dated, paste0(borg_path, "nz_borg_acc_list.txt"), quote = FALSE, sep = "\n", col.names = NA, row.names = FALSE)
ggplot(borg_dated, aes(x=Year)) +
  geom_bar() +
  theme(axis.text.x = element_text(angle=90)) #+
    scale_x_continuous(breaks = seq(2000,2020,by=2))


```

### divide interrogans isolates based on BAPs Clustering
- produced by fastGear
- can also do in fastBAPs, but prodcued more clades
```{r}
# int_path <- "/Users/rachel/Dropbox/5.Rachel-projects/Phylogeography/Lepto_information_V3/interrogans/"

int_path <- "/Users/rachel/Desktop/"
all_int_list <- read.csv(paste0(int_path, "pirate_baps_for_each_acc.csv"), sep=",", header = TRUE)
all_int_list
int_lineage_4 <- all_int_list %>% subset(pirate_baps == 5) %>% select(BioSample.Accession)

int_lineage_4
write.csv(int_lineage_4, paste0(int_path, "int_lineage_5_51.txt"), quote= FALSE, row.names = FALSE, col.names = FALSE)
```
