---
title: "all_leptopsira_WGS"
author: "Rachel Xu"
date: "1/14/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r}
library(dplyr)
path <- "/Users/rachel/Dropbox/5.Rachel-projects/Phylogeography/Lepto_information_V3/assemblies/"

```

# Read in previously organized metadata files
- merged manually to take metadata from all databases
```{r}
# # all patric wgs metadata 1/14/2021
# patric_all <- read.csv(paste0(path,"all_patric_wgs.csv")) %>% select(c("Genome.Name",	"Serovar",		"BioSample.Accession",	"Sequencing.Centers",	"Sequencing.Platform",	"Sequencing.Depth",	"GC.Content",	"Isolation.Source",	"Isolation.Country",	"Geographic.Location",	"Host.Name",	"Collection.Date",	"species",	"group",))
# 
# # current ncbi data I have extracted, but not the newly added ones (1/14/2021)
# ncbi_current <- read.csv(paste0(path, "ncbi+self_973.csv")) %>% 
#   rename(BioSample.Accession = Sample)
# ncbi_new_meta_search <- read.csv(paste0(path, "biosample_metadata_973.csv"))
# merged_ncbi <- full_join(ncbi_current, ncbi_new_meta_search, by="BioSample.Accession")
# 
# # read in sapelo2 samples that were not in V2 assemblies' metadata
# sapelo2_extra <- read.csv(paste0(path, "biosample_metadata_sapelo.csv"))
# 
# # merge metadata together, also to see which biosample still need metadata 
# merged_metadata <- full_join(patric_all, merged_ncbi, by= "BioSample.Accession")
# merged_metadata <- full_join(merged_metadata, sapelo2_extra, by ="BioSample.Accession")


# ****DO NOT OVERRIDE THIS FILE< MANUALLY MERGED
# write.csv(merged_metadata, paste0(path, "NCBI_patric_all_wgs_metadata.csv"))

```


# wrangling with merged metadata 

# 1) need parse out species
```{r}

all_asm_meta <- read.csv(paste0(path, "NCBI_patric_all_wgs_metadata.csv"))

# 1) need parse out species
new_species <- unlist(lapply(all_asm_meta$Genome.Name.x, function(X) {unlist(strsplit(X, " " ))[2]}))
all_asm_meta$species <- new_species
all_asm_meta <- all_asm_meta %>% select(-c("species.x"))









```

# 2) need to put species into groups
```{r}

# 2) need to put species into groups

pathogenic <- c("interrogans","mayottensis","santarosai","borgpetersenii","dzianensis","alstonii", "adleri","noguchii", "kirschneri", "barantonii", "weilii","putramalaysiae","alexanderi","gomenensis","kmetyi","yasudae","elllisii","adleri","stimsonii","kmetyi", "alstoni")

intermediate <- c("andrefontaineae","wolffii","sarikeiensis","fainei","selangorensis","inadai","dzoumogneensis","fluminis","licerasiae","koniamboensis","fletcheri","langatensis","semungkisensis","perolatii","neocaledonica","haakeii","saintgironsiae","hartskeerlii","johnsonii","venezuelensis", "broomii")

saprophytic <- c("meyeri", "idonii", "wolbachii","biflexa", "levettii", "brenneri", "ilyithenensis", "bandrabouensis", "bouyouniensis", "mtsangambouensis", "jelokensis", "congkakensis", "kemamanensis", "ognonensis", "perdikensis", "montravelensis", "bourretii", "noumeaensis", "kanakyensis", "vanthielii")

# label all lepto genomes from PATRIC with phylogenetic group              
all_asm_meta <- all_asm_meta %>% 
  subset(species != "illini") %>% 
  mutate(group=ifelse(species %in% pathogenic, "pathogenic", 
                      ifelse(species %in% intermediate, "intermediate", 
                             ifelse(species %in% saprophytic, "saprophytic", "unknown")))) 

```


# parse serovars
```{r}

all_asm_meta <- all_asm_meta %>%
  mutate(Serovar=tolower(ifelse(grepl("serovar",Genome.Name.x,fixed = TRUE), unlist(lapply(Genome.Name.x,function(x) {unlist(strsplit(x," "))[match("serovar", unlist(strsplit(x," ")))+1]})), Serovar.x)))  

```

### Serogroup dictionary
```{r}

Anarum <- c("Pingchang")
Andaman <- c("Andamana")
Australis <- c("Australis","Bajan","Bangkok","Bratislava","Fugis","Hawain","Jalna","Lora","Muenchen","Nicaragua","Peruviana","Pina","Ramisi","Rushan","Soteropolitana")
Autumnalis <- c("Alice","Autumnalis","Bangkinang","Bim","Bulgarica","Butembo","Carlos","Erinaceiauriti","Fortbragg","Lambwe","Mooris","Mujunkumi","Nanla","Rachmati","Srebarna","Sumatrana","Weerasinghe")
Ballum <- c("Aroborea","Ballum","Castellonis","Guangdong","Kenya","Peru","Sichuan")
Bataviae <- c("Argentiniensis","Balboa","Bataviae","Brasiliensis","Claytoni","Djatzi","Kobbe","Losbanos","Paidjan","Rioja","Santarosa","Waimea")
Canicola <- c("Bafani","Benjamini","Bindjei","Broomi","Canicola","Galtoni","Jonsis","Kamituga","Kuwait","Malaya","Manilae","Portlandvere","Qunjian","Schueffneri","Sumneri")
Celledoni <- c("Anhoa","Celledoni","Cynopteri","Hainan","Mengdeng","Whitcombi")
Codice<- c("Codice")
Cynopteri <- c("Cynopteri","Tingomaria")
Djasiman <- c("Agogo","Buenos Aires","Djasiman","Gurungi","Huallaga","Sentot")
Grippotyphosa <- c("Bananal","Canalzonae","Dadas","Grippotyphosa","Grippotyphosa Type Duyster","Grippotyphosa Type Moskva","Honghe","Huanuco","Liangguang","Linhai","Muelleri","Ratnapura","Valbuzzi","Vanderhoedeni")
Hebdomanis <- c("Borincana","Goiano","Hebdomadis","Jules","Kabura","Kambale","Kremastos","Manzhuang","Maru","Medanensis","Nanding","Nona","Sanmartini","Worsfoldi")
Holland <- c("Holland")
Hurstbridge <- c("Hurstbridge","Hurstbridge type HB6")
Icterohaemorrhagiae <- c("Birkini","Bogvere","Copenhageni","Dakota","Gem","Hongchon","Honghe","Hualin","Icterohaemorrhagiae","Lai","Lai Type Langkawi","Liangshan","Mankarso","Monymusk","Mwogolo","Naam","Nanxi","Ndahambukuje","Ndambari","Smithi","Sokoine","Tonkini","Yeoncheon", "copenhageni/icterohaemorrhagiae")
Iquitos <- c("Varillal")
Javanica <- c("Arenal","Ceylonica","Coxi","Dehong","Fluminense","Javanica","Kalimantani","Mengla","Mengma","Mengrum","Menoni","Menrun","Poi","Sofia","Sorexjalna","Vargonicas","Yaan","Zhenkang")
Louisiana <- c("Lanka","Louisiana","Orleans")
Lyme <- c("Lyme")
Manhao <- c("Lichuan","Lincang","Lushui","Manhao","Qingshui")
Mini <- c("Beye","Georgia","Hekou","Mini","Perameles","Ruparupae","Swajizak","Tabaquite","Yunnan")
Panama <- c("Cristobali","Mangus","Panama")
Pomona <- c("Altodouro","Kennewicki","Kunming","Mozdok","Pomona","Proechimys","Tropica","Tsaratsovo")
Pyrogenes <- c("Abramis","Alexi","Biggis","Camlo","Costa Rica","Guaratuba","Hamptoni","Kwale","Manilae","Menglian","Myocastoris","Nigeria","Princestown","Pyrogenes","Robinsoni","Varela","Zanoni")
Ranarum <- c("Evansi","Pingchang","Ranarum")
Sarmin <- c("Cuica","Machiguenga","Rio","Sarmin","Waskurin","Weaveri")
Sehgali <- c("Portblairi")
Sejroe <- c("Balcanica","Caribe","Dikkeni","Geyaweera","Gorgas","Guaricura","Haemolytica","Hardjo","Hardjo type Bovis","Hardjo type Prajitno",
            "hardjo-bovis",
            "Hardjobovis",
            "hardjo-prajitno","Istrica","Jin","Medanensis","Nyanza","Polonica","Recreo","Ricardi","Roumanica","Saxkoebing","Sejroe","Trinidad","Unipertama","Wolffi")
Semaranga <- c("Patoc","Semaranga","Saopaulo")
Shermani <- c("Aguaruna","Badudieri","Carimagua","Luis","Shermani")
Tarassovi <- c("Atchafalaya","Atlantae","Bakeri","Banna"," Bravo","Chagres","Corredores","Darien","Gatuni","Gengma","Guidae","Kanana","Kaup","Kisuba","Langati","Malaysia","Mengpeng","Mogden","Moldaviae","Navet","Ningxia","Rama","Sulzerae","Tarassovi","Topaz","Tunis","Vughia","Yunxian")
Szwajizak <- c("szwajizak")
Arborea <- c("Arborea")
Heyan <- c("heyan")

serogroup_dict <- list(Anarum = Anarum,Andaman=Andaman, Australis=Australis,Autumnalis=Autumnalis,Ballum=Ballum,Bataviae=Bataviae, Canicola=Canicola,Celledoni=Celledoni,Codice=Codice,Cynopteri=Cynopteri,Djasiman=Djasiman,Grippotyphosa=Grippotyphosa,Hebdomanis=Hebdomanis,Holland=Holland,Hurstbridge=Hurstbridge,Icterohaemorrhagiae=Icterohaemorrhagiae,Iquitos=Iquitos,Javanica=Javanica,Louisiana=Louisiana,Lyme=Lyme,Manhao=Manhao,Mini=Mini,Panama=Panama,Pomona=Pomona,Pyrogenes=Pyrogenes,Ranarum=Ranarum,Sarmin=Sarmin,Sehgali=Sehgali,Sejroe=Sejroe,Semaranga=Semaranga,Shermani=Shermani,Tarassovi=Tarassovi,Szwajizak=Szwajizak, Arborea=Arborea,  Heyan = Heyan)
serogroup_dict <- lapply(serogroup_dict, tolower)


```


# 4ï¼‰assign serogroup base on serovar
```{r}
serovars <- all_asm_meta$Serovar 
serogroups <- c()
j = 0
for(item in serovars){
  j=j+1
  i=0
  in_dict <- 0
  item <- tolower(item)
  # print(paste0("serovar: ",item))
  for(serogroup in serogroup_dict){
    i=i+1
    if(item %in% serogroup){
      in_dict <- 1
      # print(paste0("serogroup: ",names(serogroup_dict)[i]))
      serogroups <- c(serogroups,names(serogroup_dict)[i])
      break
    }
  }
  if(in_dict == 0){
    # print("not_in_dict")
    serogroups <- c(serogroups,"not_in_dict")
  }
}

length(serogroups)
# serovars
all_asm_meta$Serovar_2.x <- unlist(lapply(pathogenic_sero_db$Serovar_2.x, function(x){
  ifelse(x=="Hardjobovis","hardjo-bovis",ifelse(x=="copenhageni/icterohaemorrhagiae","copenhageni",tolower(x)))
}[1]))
all_asm_meta$serogroups <-serogroups


```